### PR TITLE
APIs for Backup authenticator

### PR DESCRIPTION
**Pull request by bhagyasakalanka**
## Related PRs
merge https://github.com/wso2-extensions/identity-outbound-auth-backup-code/pull/1 before merging this PR

### CODE REVIEW

File: components/org.wso2.carbon.identity.api.user.backupcode/org.wso2.carbon.identity.api.user.backupcode.v1/src/main/java/org/wso2/carbon/identity/rest/api/user/backupcode/v1/core/BackupCodeService.java [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+import javax.ws.rs.core.Response;
+
+import static org.wso2.carbon.identity.api.user.backupcode.common.BackupCodeConstants.ErrorMessage.SERVER_ERROR_GENERAL;
+import static org.wso2.carbon.identity.api.user.backupcode.common.BackupCodeConstants.ErrorMessage.SERVER_ERROR_RETRIEVING_REALM_FOR_USER;
+import static org.wso2.carbon.identity.api.user.backupcode.common.BackupCodeConstants.ErrorMessage.SERVER_ERROR_RETRIEVING_USERSTORE_MANAGER;
+import static org.wso2.carbon.identity.api.user.backupcode.common.BackupCodeConstants.ErrorMessage.USER_ERROR_ACCESS_DENIED_FOR_BASIC_AUTH;
+import static org.wso2.carbon.identity.api.user.backupcode.common.BackupCodeConstants.ErrorMessage.USER_ERROR_UNAUTHORIZED_USER;
+
+/**
+ *
```
Comment (bhagyasakalanka) [inline components/org.wso2.carbon.identity.api.user.backupcode/org.wso2.carbon.identity.api.user.backupcode.v1/src/main/java/org/wso2/carbon/identity/rest/api/user/backupcode/v1/core/BackupCodeService.java:53]: Add the description

File: components/org.wso2.carbon.identity.api.user.backupcode/org.wso2.carbon.identity.api.user.backupcode.v1/src/main/java/org/wso2/carbon/identity/rest/api/user/backupcode/v1/core/BackupCodeService.java [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+ */
+public class BackupCodeService {
+
+    private static final Log log = LogFactory.getLog(BackupCodeService.class);
+
```
Comment (bhagyasakalanka) [inline components/org.wso2.carbon.identity.api.user.backupcode/org.wso2.carbon.identity.api.user.backupcode.v1/src/main/java/org/wso2/carbon/identity/rest/api/user/backupcode/v1/core/BackupCodeService.java:58]: Add method comment

File: components/org.wso2.carbon.identity.api.user.backupcode/org.wso2.carbon.identity.api.user.backupcode.v1/src/main/java/org/wso2/carbon/identity/rest/api/user/backupcode/v1/core/BackupCodeService.java [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+     * Get authenticated user.
+     *
+     * @return Authenticated user
+     */
+    public static User getUser() {
+
+        return ContextLoader.getUserFromContext();
+    }
+
+    private boolean isValidAuthenticationType() {
```
Comment (bhagyasakalanka) [inline components/org.wso2.carbon.identity.api.user.backupcode/org.wso2.carbon.identity.api.user.backupcode.v1/src/main/java/org/wso2/carbon/identity/rest/api/user/backupcode/v1/core/BackupCodeService.java:129]: Verify the logic of this method

File: components/org.wso2.carbon.identity.api.user.backupcode/org.wso2.carbon.identity.api.user.backupcode.v1/src/main/java/org/wso2/carbon/identity/rest/api/user/backupcode/v1/core/BackupCodeService.java [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+            throw handleError(Response.Status.FORBIDDEN, USER_ERROR_ACCESS_DENIED_FOR_BASIC_AUTH);
+        }
+
+        UserRealm userRealm;
+        String backupCodes = null;
+        User user = getUser();
+        String tenantAwareUsername = null;
+
+        try {
+            userRealm = BackupCodeUtil.getUserRealm(user.toFullQualifiedUsername());
```
Comment (bhagyasakalanka) [inline components/org.wso2.carbon.identity.api.user.backupcode/org.wso2.carbon.identity.api.user.backupcode.v1/src/main/java/org/wso2/carbon/identity/rest/api/user/backupcode/v1/core/BackupCodeService.java:218]: Check the possibility of getting user store manager from a private method

File: components/org.wso2.carbon.identity.api.user.backupcode/org.wso2.carbon.identity.api.user.backupcode.v1/src/main/java/org/wso2/carbon/identity/rest/api/user/backupcode/v1/core/BackupCodeService.java (hunk 3 above)
Comment (bhagyasakalanka) [reply components/org.wso2.carbon.identity.api.user.backupcode/org.wso2.carbon.identity.api.user.backupcode.v1/src/main/java/org/wso2/carbon/identity/rest/api/user/backupcode/v1/core/BackupCodeService.java:138]: The logic here is valid. The below comment explain the requirement of this validation
```
        /*
         * Check whether the request is authenticated with basic auth. Backup code endpoint should not be allowed for
         * basic authentication. This approach can be improved by providing a Level of Assurance (LOA) and
         * checking that in BackupCodeService.
         */
```

File: components/org.wso2.carbon.identity.api.user.backupcode/org.wso2.carbon.identity.api.user.backupcode.common/src/main/java/org/wso2/carbon/identity/api/user/backupcode/common/BackupCodeConstants.java [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+package org.wso2.carbon.identity.api.user.backupcode.common;
+
+/**
+ * Contains all the Backup Code related constants
+ */
+public class BackupCodeConstants {
+
+    public static final String USER_BACKUP_CODE_PREFIX = "BCA-";
+
```
Comment (somindatommy) [inline components/org.wso2.carbon.identity.api.user.backupcode/org.wso2.carbon.identity.api.user.backupcode.common/src/main/java/org/wso2/carbon/identity/api/user/backupcode/common/BackupCodeConstants.java:25]: Lets remove this unwanted new line

File: components/org.wso2.carbon.identity.api.user.backupcode/org.wso2.carbon.identity.api.user.backupcode.v1/src/gen/java/org/wso2/carbon/identity/rest/api/user/backupcode/v1/MeApiService.java [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+import org.wso2.carbon.identity.rest.api.user.backupcode.v1.dto.UserRequestDTO;
+
+import java.util.List;
+
+import java.io.InputStream;
+import org.apache.cxf.jaxrs.ext.multipart.Attachment;
+
+import javax.ws.rs.core.Response;
+
+public abstract class MeApiService {
```
Comment (JKAUSHALYA) [inline components/org.wso2.carbon.identity.api.user.backupcode/org.wso2.carbon.identity.api.user.backupcode.v1/src/gen/java/org/wso2/carbon/identity/rest/api/user/backupcode/v1/MeApiService.java:33]: Can we add class level and method level doc comments for this class?

File: components/org.wso2.carbon.identity.api.user.backupcode/org.wso2.carbon.identity.api.user.backupcode.v1/src/gen/java/org/wso2/carbon/identity/rest/api/user/backupcode/v1/factories/MeApiServiceFactory.java [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+
+import org.wso2.carbon.identity.rest.api.user.backupcode.v1.MeApiService;
+import org.wso2.carbon.identity.rest.api.user.backupcode.v1.impl.MeApiServiceImpl;
+
+public class MeApiServiceFactory {
+
+   private final static MeApiService service = new MeApiServiceImpl();
+
+   public static MeApiService getMeApi()
+   {
```
Comment (JKAUSHALYA) [inline components/org.wso2.carbon.identity.api.user.backupcode/org.wso2.carbon.identity.api.user.backupcode.v1/src/gen/java/org/wso2/carbon/identity/rest/api/user/backupcode/v1/factories/MeApiServiceFactory.java:27]: Formatting issue.

File: components/org.wso2.carbon.identity.api.user.backupcode/org.wso2.carbon.identity.api.user.backupcode.v1/src/gen/java/org/wso2/carbon/identity/rest/api/user/backupcode/v1/MeApiService.java (hunk 6 above)
Comment (bhagyasakalanka) [reply components/org.wso2.carbon.identity.api.user.backupcode/org.wso2.carbon.identity.api.user.backupcode.v1/src/gen/java/org/wso2/carbon/identity/rest/api/user/backupcode/v1/MeApiService.java:33]: These are auto generated class. So IMO we should not change these class

File: components/org.wso2.carbon.identity.api.user.backupcode/org.wso2.carbon.identity.api.user.backupcode.v1/src/gen/java/org/wso2/carbon/identity/rest/api/user/backupcode/v1/factories/MeApiServiceFactory.java (hunk 7 above)
Comment (bhagyasakalanka) [reply components/org.wso2.carbon.identity.api.user.backupcode/org.wso2.carbon.identity.api.user.backupcode.v1/src/gen/java/org/wso2/carbon/identity/rest/api/user/backupcode/v1/factories/MeApiServiceFactory.java:27]: These are auto generated classes. So IMO we should not change these classes

File: components/org.wso2.carbon.identity.api.user.backupcode/org.wso2.carbon.identity.api.user.backupcode.v1/src/gen/java/org/wso2/carbon/identity/rest/api/user/backupcode/v1/MeApi.java [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+
+import java.util.List;
+
+import java.io.InputStream;
+import org.apache.cxf.jaxrs.ext.multipart.Attachment;
+import org.apache.cxf.jaxrs.ext.multipart.Multipart;
+
+import javax.validation.Valid;
+import javax.ws.rs.core.Response;
+import javax.ws.rs.*;
```
Comment (Yasara-Y) [inline components/org.wso2.carbon.identity.api.user.backupcode/org.wso2.carbon.identity.api.user.backupcode.v1/src/gen/java/org/wso2/carbon/identity/rest/api/user/backupcode/v1/MeApi.java:38]: Can we get rid of * imports and import only the required ones?

File: components/org.wso2.carbon.identity.api.user.backupcode/org.wso2.carbon.identity.api.user.backupcode.v1/src/gen/java/org/wso2/carbon/identity/rest/api/user/backupcode/v1/MeApi.java [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+ */
+
+package org.wso2.carbon.identity.rest.api.user.backupcode.v1;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.wso2.carbon.identity.rest.api.user.backupcode.v1.dto.*;
+import org.wso2.carbon.identity.rest.api.user.backupcode.v1.MeApiService;
+import org.wso2.carbon.identity.rest.api.user.backupcode.v1.factories.MeApiServiceFactory;
+
+import io.swagger.annotations.ApiParam;
```
Comment (Yasara-Y) [inline components/org.wso2.carbon.identity.api.user.backupcode/org.wso2.carbon.identity.api.user.backupcode.v1/src/gen/java/org/wso2/carbon/identity/rest/api/user/backupcode/v1/MeApi.java:24]: Shall we re-arrange these imports in a methodical way?

File: components/org.wso2.carbon.identity.api.user.backupcode/org.wso2.carbon.identity.api.user.backupcode.v1/src/gen/java/org/wso2/carbon/identity/rest/api/user/backupcode/v1/MeApi.java [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+        
+        @io.swagger.annotations.ApiResponse(code = 401, message = "Unauthorized"),
+        
+        @io.swagger.annotations.ApiResponse(code = 500, message = "Server Error") })
+
+    public Response meBackupCodePost(@ApiParam(value = "Actions supported by the API. Actions can be INIT or REFRESH" ,required=true ) @Valid UserRequestDTO request) {
+
+        return delegate.meBackupCodePost(request);
+    }
+
```
Comment (Yasara-Y) [inline components/org.wso2.carbon.identity.api.user.backupcode/org.wso2.carbon.identity.api.user.backupcode.v1/src/gen/java/org/wso2/carbon/identity/rest/api/user/backupcode/v1/MeApi.java:112]: extra line

File: components/org.wso2.carbon.identity.api.user.backupcode/org.wso2.carbon.identity.api.user.backupcode.v1/src/gen/java/org/wso2/carbon/identity/rest/api/user/backupcode/v1/dto/BackupCodeResponseDTO.java [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.wso2.carbon.identity.rest.api.user.backupcode.v1.dto;
+
+import java.util.ArrayList;
+import java.util.List;
+import io.swagger.annotations.*;
```
Comment (Yasara-Y) [inline components/org.wso2.carbon.identity.api.user.backupcode/org.wso2.carbon.identity.api.user.backupcode.v1/src/gen/java/org/wso2/carbon/identity/rest/api/user/backupcode/v1/dto/BackupCodeResponseDTO.java:21]: Avoid * imports. Check other places too.

File: components/org.wso2.carbon.identity.api.user.backupcode/org.wso2.carbon.identity.api.user.backupcode.v1/src/gen/java/org/wso2/carbon/identity/rest/api/user/backupcode/v1/dto/BackupCodeResponseDTO.java [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+    @Valid 
+    private List<String> backupCodes = new ArrayList<String>();
+
+    /**
+    * Backup codes of the authenticated user
+    **/
+    @ApiModelProperty(value = "Backup codes of the authenticated user")
+    @JsonProperty("backupCodes")
+    public List<String> getBackupCodes() {
+        return backupCodes;
```
Comment (Yasara-Y) [inline components/org.wso2.carbon.identity.api.user.backupcode/org.wso2.carbon.identity.api.user.backupcode.v1/src/gen/java/org/wso2/carbon/identity/rest/api/user/backupcode/v1/dto/BackupCodeResponseDTO.java:40]: Lets add a new line at the beginning of each method

File: components/org.wso2.carbon.identity.api.user.backupcode/org.wso2.carbon.identity.api.user.backupcode.v1/src/gen/java/org/wso2/carbon/identity/rest/api/user/backupcode/v1/dto/BackupCodeResponseDTO.java [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+    }
+    public void setBackupCodes(List<String> backupCodes) {
+        this.backupCodes = backupCodes;
```
Comment (Yasara-Y) [inline components/org.wso2.carbon.identity.api.user.backupcode/org.wso2.carbon.identity.api.user.backupcode.v1/src/gen/java/org/wso2/carbon/identity/rest/api/user/backupcode/v1/dto/BackupCodeResponseDTO.java:43]: Add a new line

File: components/org.wso2.carbon.identity.api.user.backupcode/org.wso2.carbon.identity.api.user.backupcode.v1/src/gen/java/org/wso2/carbon/identity/rest/api/user/backupcode/v1/dto/ErrorDTO.java [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+    @NotNull(message = "Property message cannot be null.") 
+    private String message = null;
+
+    @Valid 
+    private String description = null;
+
+    @Valid 
+    private String traceId = null;
+
+    /**
```
Comment (Yasara-Y) [inline components/org.wso2.carbon.identity.api.user.backupcode/org.wso2.carbon.identity.api.user.backupcode.v1/src/gen/java/org/wso2/carbon/identity/rest/api/user/backupcode/v1/dto/ErrorDTO.java:43]: Incomplete doc comment

File: components/org.wso2.carbon.identity.api.user.backupcode/org.wso2.carbon.identity.api.user.backupcode.v1/src/gen/java/org/wso2/carbon/identity/rest/api/user/backupcode/v1/MeApi.java (hunk 8 above)
Comment (bhagyasakalanka) [reply components/org.wso2.carbon.identity.api.user.backupcode/org.wso2.carbon.identity.api.user.backupcode.v1/src/gen/java/org/wso2/carbon/identity/rest/api/user/backupcode/v1/MeApi.java:38]: These are auto generated classes. AFAIK, we are not modifying these classes

File: components/org.wso2.carbon.identity.api.user.backupcode/org.wso2.carbon.identity.api.user.backupcode.v1/src/gen/java/org/wso2/carbon/identity/rest/api/user/backupcode/v1/dto/ErrorDTO.java [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+    **/
+    @ApiModelProperty(value = "")
+    @JsonProperty("traceId")
+    public String getTraceId() {
+        return traceId;
+    }
+    public void setTraceId(String traceId) {
+        this.traceId = traceId;
```
Comment (Yasara-Y) [inline components/org.wso2.carbon.identity.api.user.backupcode/org.wso2.carbon.identity.api.user.backupcode.v1/src/gen/java/org/wso2/carbon/identity/rest/api/user/backupcode/v1/dto/ErrorDTO.java:84]: Add a new line. Check other places too

File: components/org.wso2.carbon.identity.api.user.backupcode/org.wso2.carbon.identity.api.user.backupcode.v1/src/gen/java/org/wso2/carbon/identity/rest/api/user/backupcode/v1/MeApi.java (hunk 9 above)
Comment (bhagyasakalanka) [reply components/org.wso2.carbon.identity.api.user.backupcode/org.wso2.carbon.identity.api.user.backupcode.v1/src/gen/java/org/wso2/carbon/identity/rest/api/user/backupcode/v1/MeApi.java:24]: These are auto generated classes. AFAIK, we are not modifying these classes

File: components/org.wso2.carbon.identity.api.user.backupcode/org.wso2.carbon.identity.api.user.backupcode.v1/src/gen/java/org/wso2/carbon/identity/rest/api/user/backupcode/v1/factories/MeApiServiceFactory.java (hunk 7 above)
Comment (Yasara-Y) [inline components/org.wso2.carbon.identity.api.user.backupcode/org.wso2.carbon.identity.api.user.backupcode.v1/src/gen/java/org/wso2/carbon/identity/rest/api/user/backupcode/v1/factories/MeApiServiceFactory.java:27]: Formatting issue

File: components/org.wso2.carbon.identity.api.user.backupcode/org.wso2.carbon.identity.api.user.backupcode.v1/src/main/java/org/wso2/carbon/identity/rest/api/user/backupcode/v1/core/BackupCodeService.java [hunk 16]
```diff
@@ -0,0 +1,232 @@
...
+        }
+
+        BackupCodeResponseDTO backupCodeResponseDTO = new BackupCodeResponseDTO();
+        backupCodeResponseDTO.setBackupCodes(backupCodes);
+
+        return backupCodeResponseDTO;
+    }
+
+    /**
+     * Initialized backup codes
```
Comment (Yasara-Y) [inline components/org.wso2.carbon.identity.api.user.backupcode/org.wso2.carbon.identity.api.user.backupcode.v1/src/main/java/org/wso2/carbon/identity/rest/api/user/backupcode/v1/core/BackupCodeService.java:77]: Missing the fullstop.

File: components/org.wso2.carbon.identity.api.user.backupcode/org.wso2.carbon.identity.api.user.backupcode.v1/src/main/java/org/wso2/carbon/identity/rest/api/user/backupcode/v1/core/BackupCodeService.java [hunk 17]
```diff
@@ -0,0 +1,232 @@
...
+            BackupCodeAPIHandler.deleteBackupCodes(user.toFullQualifiedUsername());
+        } catch (BackupCodeException ex) {
+            throw handleException(ex, SERVER_ERROR_DELETING_BACKUP_CODES);
+        }
+    }
+
+    /**
+     * Get authenticated user.
+     *
+     * @return Authenticated user
```
Comment (Yasara-Y) [inline components/org.wso2.carbon.identity.api.user.backupcode/org.wso2.carbon.identity.api.user.backupcode.v1/src/main/java/org/wso2/carbon/identity/rest/api/user/backupcode/v1/core/BackupCodeService.java:155]: Add a fullstop.

File: components/org.wso2.carbon.identity.api.user.backupcode/org.wso2.carbon.identity.api.user.backupcode.v1/src/gen/java/org/wso2/carbon/identity/rest/api/user/backupcode/v1/factories/MeApiServiceFactory.java (hunk 7 above)
Comment (bhagyasakalanka) [reply components/org.wso2.carbon.identity.api.user.backupcode/org.wso2.carbon.identity.api.user.backupcode.v1/src/gen/java/org/wso2/carbon/identity/rest/api/user/backupcode/v1/factories/MeApiServiceFactory.java]: fixed formatting issue

File: components/org.wso2.carbon.identity.api.user.backupcode/org.wso2.carbon.identity.api.user.backupcode.v1/src/gen/java/org/wso2/carbon/identity/rest/api/user/backupcode/v1/MeApiService.java (hunk 6 above)
Comment (bhagyasakalanka) [reply components/org.wso2.carbon.identity.api.user.backupcode/org.wso2.carbon.identity.api.user.backupcode.v1/src/gen/java/org/wso2/carbon/identity/rest/api/user/backupcode/v1/MeApiService.java:36]: Added method and class level comments

File: components/org.wso2.carbon.identity.api.user.backupcode/org.wso2.carbon.identity.api.user.backupcode.v1/src/main/java/org/wso2/carbon/identity/rest/api/user/backupcode/v1/core/BackupCodeService.java (hunk 16 above)
Comment (bhagyasakalanka) [reply components/org.wso2.carbon.identity.api.user.backupcode/org.wso2.carbon.identity.api.user.backupcode.v1/src/main/java/org/wso2/carbon/identity/rest/api/user/backupcode/v1/core/BackupCodeService.java]: Fixed

File: components/org.wso2.carbon.identity.api.user.backupcode/org.wso2.carbon.identity.api.user.backupcode.v1/src/main/java/org/wso2/carbon/identity/rest/api/user/backupcode/v1/core/BackupCodeService.java (hunk 17 above)
Comment (bhagyasakalanka) [reply components/org.wso2.carbon.identity.api.user.backupcode/org.wso2.carbon.identity.api.user.backupcode.v1/src/main/java/org/wso2/carbon/identity/rest/api/user/backupcode/v1/core/BackupCodeService.java]: fixed

File: components/org.wso2.carbon.identity.api.user.backupcode/org.wso2.carbon.identity.api.user.backupcode.v1/src/gen/java/org/wso2/carbon/identity/rest/api/user/backupcode/v1/factories/MeApiServiceFactory.java (hunk 7 above)
Comment (bhagyasakalanka) [reply components/org.wso2.carbon.identity.api.user.backupcode/org.wso2.carbon.identity.api.user.backupcode.v1/src/gen/java/org/wso2/carbon/identity/rest/api/user/backupcode/v1/factories/MeApiServiceFactory.java]: fixed

**Review by Yasara-Y — APPROVED** [review]

**Pull request merged**